### PR TITLE
RHCLOUD-28805 | feature: tune redelivery conditions for the HTTP connectors

### PR DESCRIPTION
**Pull request by MikelAlejoBR**
With certain responses from the external services, we want the HTTP connectors to automatically retry the redelivery.

## Jira ticket
[[RHCLOUD-28805]](https://issues.redhat.com/browse/RHCLOUD-28805)

### CODE REVIEW

File: connector-common-http/pom.xml [hunk 1]
```diff
@@ -59,6 +59,18 @@
...
+        <dependency>
+            <groupId>org.apache.camel.quarkus</groupId>
+            <artifactId>camel-quarkus-junit5</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <scope>test</scope>
+        </dependency>
```
Comment (gwenneg) [inline connector-common-http/pom.xml:72]:
```suggestion
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-junit5-mockito</artifactId>
            <scope>test</scope>
        </dependency>
```

File: connector-common-http/src/main/java/com/redhat/cloud/notifications/connector/http/HttpRedeliveryPredicate.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+        if (t instanceof HttpOperationFailedException e) {
+            final boolean shouldRetry = e.getStatusCode() >= 500 || e.getStatusCode() == HttpStatus.SC_TOO_MANY_REQUESTS;
+
+            Log.debugf("The HTTP request failed with status code '%s' and body '%s'. %s", e.getStatusCode(), e.getResponseBody(), (shouldRetry) ? "Retrying..." : "Not retrying");
+
+            return shouldRetry;
+        }
+
+        return t instanceof SocketTimeoutException;
```
Comment (gwenneg) [inline connector-common-http/src/main/java/com/redhat/cloud/notifications/connector/http/HttpRedeliveryPredicate.java]: Is there any reason why you favored `SocketTimeoutException` over `IOException` here?
```suggestion
        return t instanceof IOException;
```

Comment (gwenneg) [reply connector-common-http/src/main/java/com/redhat/cloud/notifications/connector/http/HttpRedeliveryPredicate.java]: I commented a previous version of the code. The question still stands with `NoHttpResponseException` and `SocketTimeoutException` which both extend `IOException`.

Comment (MikelAlejoBR) [reply connector-common-http/src/main/java/com/redhat/cloud/notifications/connector/http/HttpRedeliveryPredicate.java]: Not really, I thought we needed to be more specific, but I'm good with having `IOException`. Let me change it

File: connector-common-http/src/test/java/com/redhat/cloud/notifications/connector/http/HttpRedeliveryPredicateTest.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+     * Tests that the redelivery predicate returns {@code true} for when an
+     * IOException is thrown.
+     */
+    @Test
+    void testPositiveMatchIOException() {
+        // Simulate a timeout exception.
+        final Exchange exchange = this.createExchangeWithBody("");
+
+        final SocketTimeoutException exception = Mockito.mock(SocketTimeoutException.class);
+        exchange.setProperty(Exchange.EXCEPTION_CAUGHT, exception);
```
Comment (gwenneg) [inline connector-common-http/src/test/java/com/redhat/cloud/notifications/connector/http/HttpRedeliveryPredicateTest.java:67]:
```suggestion
        exchange.setProperty(Exchange.EXCEPTION_CAUGHT, new SocketTimeoutException());
```

File: connector-common/src/test/java/com/redhat/cloud/notifications/connector/ConnectorRoutesTest.java [hunk 4]
```diff
@@ -125,17 +125,26 @@ protected void testSuccessfulNotification() throws Exception {
...
+        testFailedNotification(this.connectorConfig.getRedeliveryMaxAttempts());
     }
 
     @Test
     protected void testFailedNotificationError404() throws Exception {
         mockRemoteServerError(404, "Page not found");
-        testFailedNotification();
+
+        // We expect the connector to not retry the notification since the
+        // mocked request returns a 400.
```
Comment (gwenneg) [inline connector-common/src/test/java/com/redhat/cloud/notifications/connector/ConnectorRoutesTest.java:142]:
```suggestion
        // mocked request returns a 404.
```

**Comment by gwenneg** [comment]
Sonar was unhappy because of some missing Kafka config so I copied and slightly changed `application.properties` from `notifications-connector-common`.

**Pull request merged**